KAFKA-13436: Omitted BrokerTopicMetrics metrics in the documentation

dongjinleekr opened the pull request:
I found this issue while working on [KAFKA-13354](https://issues.apache.org/jira/browse/KAFKA-13354).

As soon as this PR is approved, I will open a PR on the kafka-site also.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

dongjinleekr committed:
Add omitted BrokerTopicMetrics metrics to the documentation

dongjinleekr commented:
@splett2 Could you have a look while I am updating #11474?

splett2 commented on docs/ops.html:
```diff
@@ -1447,12 +1447,32 @@ <h4 class="anchor-heading"><a id="remote_jmx" class="anchor-link"></a><a href="#
+      <tr>
+        <td>Failed produce request rate</td>
+        <td>kafka.server:type=BrokerTopicMetrics,name=FailedProduceRequestsPerSec</td>
+        <td></td>
```
seems like we're missing an explanation for what these metrics measure.

mimaison replied:
Yes let's try to include descriptions as above we recommend to graph and alert on these metrics

dongjinleekr committed:
1. Add descriptions to the BrokerTopicMetrics metrics. 2. Add ',topic=([-.\w]+)' to the BrokerTopicMetrics metrics.

dongjinleekr commented:
@splett2 Here is the fix. Reviewing the documentation, I found that 1. the other 'BrokerTopicMetrics' metrics are omitting `,topic=([-.\w]+)` at the end 2. also omitting the descriptions. I fixed all of them at once.

dongjinleekr commented:
As soon as this PR is merged, I will open a corresponding PR in `kafka-site`.

dajac commented on docs/ops.html:
```diff
@@ -1411,18 +1411,18 @@ <h4 class="anchor-heading"><a id="remote_jmx" class="anchor-link"></a><a href="#
-        <td>kafka.server:type=BrokerTopicMetrics,name=MessagesInPerSec</td>
-        <td></td>
+        <td>kafka.server:type=BrokerTopicMetrics,name=MessagesInPerSec,topic=([-.\w]+)</td>
+        <td>Number of incoming messages per topic.</td>
```
The third column's header is `Normal value` so it feels a bit weird to put a description in there. The description is actually the first column. Perhaps, we could complement the description when necessary instead of adding this?

dongjinleekr committed:
Remove repetitive descriptions.

dongjinleekr committed:
Recover 'Message conversion rate' description

dongjinleekr commented:
@dajac Here is the update; As you can see here, only `Message conversion rate` and `Rejected byte rate` have detailed descriptions now.

dongjinleekr committed:
1. Reorder Error rate metric. 2. Added descriptions on newly added BrokerTopicMetrics metrics.

dongjinleekr committed:
Update documentation on BrokerTopicMetrics metrics

1. Add description on all-topic metrics; these metrics are provided on both of per-topic, all-topic basis. (see: BrokerTopicStats#topicStats, BrokerTopicStats#allTopicsStats)

  - MessagesInPerSec
  - BytesInPerSec
  - BytesOutPerSec
  - ReplicationBytesInPerSec
  - ReplicationBytesOutPerSec
  - TotalProduceRequestsPerSec
  - TotalFetchRequestsPerSec
  - FailedProduceRequestsPerSec
  - FailedFetchRequestsPerSec
  - ProduceMessageConversionsPerSec
  - FetchMessageConversionsPerSec
  - BytesRejectedPerSec

2. Remove the 'topic' attribute from all-topic only metrics:

  - NoKeyCompactedTopicRecordsPerSec
  - InvalidMagicNumberRecordsPerSec
  - InvalidMessageCrcRecordsPerSec
  - InvalidOffsetOrSequenceRecordsPerSec
  - ReassignmentBytesOutPerSec
  - ReassignmentBytesInPerSec

dongjinleekr committed:
Add normal value descriptions for all-topic basis only metrics

dongjinleekr commented:
@splett2 @mimaison Sorry for being late. Here is the fix.

Not only rebasing onto the latest trunk, I carefully inspected the implementation of `BrokerTopicStats` and did the following:

- For the metrics provided on both of per-topic and all-topic basis (including newly documented four), the `topic=([-.\w]+)` tag and detailed descriptions are added, with an explanation on all-topic metrics.
- For the metrics provided on an all-topic basis only, I added some simple descriptions on normal values for consistency.

dongjinleekr commented:
Retest this please.

dongjinleekr commented:
Rebased onto the latest trunk.

mimaison commented on docs/ops.html:
```diff
@@ -1440,7 +1440,27 @@ <h4 class="anchor-heading"><a id="remote_jmx" class="anchor-link"></a><a href="#
+      <tr>
+        <td>Fetch request rate</td>
+        <td>kafka.server:type=BrokerTopicMetrics,name=TotalFetchRequestsPerSec,topic=([-.\w]+)</td>
+        <td>Fetch request (from client or follower) rate per topic. Omitting 'topic=(...)' will yield the all-topic rate.</td>
```
`from client or follower` -> `from clients or followers`. Same below

mimaison commented on docs/ops.html:
```diff
@@ -1460,7 +1480,7 @@ <h4 class="anchor-heading"><a id="remote_jmx" class="anchor-link"></a><a href="#
         <td>Message conversion rate</td>
         <td>kafka.server:type=BrokerTopicMetrics,name={Produce|Fetch}MessageConversionsPerSec,topic=([-.\w]+)</td>
-        <td>Number of records which required message format conversion.</td>
+        <td>Message format conversion rate, by Produce or Fetch request, per topic. Omitting 'topic=(...)' will yield the all-topic rate.</td>
```
`by Produce or Fetch request` -> `for Produce or Fetch requests`

mimaison commented on docs/ops.html:
```diff
@@ -1469,33 +1489,38 @@ <h4 class="anchor-heading"><a id="remote_jmx" class="anchor-link"></a><a href="#
+      <tr>
+        <td>Rejected byte rate</td>
+        <td>kafka.server:type=BrokerTopicMetrics,name=BytesRejectedPerSec,topic=([-.\w]+)</td>
+        <td>Rejected byte rate per topic, due to the record batch size is greater than max.message.bytes configuration. Omitting 'topic=(...)' will yield the all-topic rate.</td>
```
`is greater` -> `being greater`

mimaison commented on docs/ops.html:
```diff
@@ -1739,12 +1764,12 @@ <h4 class="anchor-heading"><a id="remote_jmx" class="anchor-link"></a><a href="#
         <td>Outgoing byte rate of reassignment traffic</td>
         <td>kafka.server:type=BrokerTopicMetrics,name=ReassignmentBytesOutPerSec</td>
-        <td></td>
+        <td>0; non-zero when the partition reassignment is in progress.</td>
```
`the partition reassignment` -> `a partition reassignment`
Same below

mimaison reviewed (commented):
Thanks @dongjinleekr for the PR. I left a few suggestions

dongjinleekr committed:
Fix grammatical errors (thanks to @mimaison)

dongjinleekr commented:
@mimaison Here it is, with rebasing onto the latest trunk. :bow:

mimaison reviewed (approved):
LGTM
